### PR TITLE
[SPARK-49249][SPARK-49122][Connect][SQL] Add `addArtifact` API to the Spark SQL Core

### PR DESCRIPTION
**Pull request by xupefei**
### What changes were proposed in this pull request?

This PR improves Spark SQL Core by adding a bunch of `addArtifact` APIs to `SparkSession`. These APIs were first introduced to Spark Connect a while ago.

The follow-up task is for PySpark.

### Why are the changes needed?

To close the API compatibility gap between Spark Connect and Spark Classic.

### Does this PR introduce _any_ user-facing change?

Yes, users will be able to use some new APIs.

### How was this patch tested?

Added new tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -900,6 +988,29 @@ class SparkSession private(
...
+        val tempFile = tempDir.resolve(artifact.path.getFileName)
+        val outStream = Files.newOutputStream(tempFile)
+        Utils.tryWithSafeFinallyAndFailureCallbacks {
+          d.stream.transferTo(outStream)
+          artifactManager.addArtifact(artifact.path, tempFile, fragment = None)
+        }(finallyBlock = {
+          outStream.close()
+        })
+      case _ =>
+        throw new IllegalArgumentException(s"Unsupported artifact storage: ${artifact.storage}")
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:1010]: Use `SparkException.internalError(..)`?

File: sql/api/src/main/scala/org/apache/spark/sql/Artifact.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import java.nio.file.{Files, Path, Paths}
+
+import org.apache.commons.lang3.StringUtils
+
+import org.apache.spark.sql.Artifact.LocalData
+import org.apache.spark.util.ArrayImplicits._
+import org.apache.spark.util.MavenUtils
+
+
+class Artifact private(val path: Path, val storage: LocalData) {
```
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/Artifact.scala:31]: This should not be public API.

Comment (hvanhovell) [reply sql/api/src/main/scala/org/apache/spark/sql/Artifact.scala:31]: Can this be unified with the client class?

**Review by hvanhovell — COMMENTED** [review]
Look pretty good!

**Comment by hvanhovell** [comment]
@xupefei can you check if this works with UDFs?

File: sql/api/src/main/scala/org/apache/spark/sql/Artifact.scala (hunk 2 above)
Comment (xupefei) [reply sql/api/src/main/scala/org/apache/spark/sql/Artifact.scala:31]: Done!

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -900,6 +988,29 @@ class SparkSession private(
     internalCreateDataFrame(rowRdd, schema)
   }
 
+  private def addArtifactsInternal(artifacts: Seq[Artifact]): Unit = artifacts.foreach { artifact =>
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:991]: Can you put this in ArtifactManager?

**Review by hvanhovell — APPROVED** [review]
One small nit...

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 3 above)
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:991]: Done!

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by hvanhovell — APPROVED** [review]

**Pull request closed without merging**